평균 제곱근 오차(Root Mean Squared Error, RMSE)  

임의의 선을 그리고 난 후 이 선이 얼마나 잘 그려졌는지 평가해서 조금씩 수정해나간다.  
이를 위해서 주어진 선의 오차를 평가하는 알고리즘이 필요한데 그 중에서 가장 많이 사용되는 방법이 평균 제곱근 오차이다.

In [7]:
import warnings
warnings.filterwarnings(action='ignore')
%config Completer.use_jedi = False
import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()
import numpy as np
import pandas as pd


In [31]:
# 앞의 예제 데이터에서 [x(공부시간), y(시험성적)]
data = [[2, 81],[4, 93], [6, 91], [8, 97]]
# 기울기 a와 y절편 b를 임의로 정한다. => [기울기, y절편]
ab = [3, 76] # 2.3, 79

for i in data:
    print('공부 시간 {}, 시험 성적 {}'.format(i[0], i[1]))

공부 시간 2, 시험 성적 81
공부 시간 4, 시험 성적 93
공부 시간 6, 시험 성적 91
공부 시간 8, 시험 성적 97


In [32]:
# x, y의 데이터 값이 1건씩 저장된 리스트 data에서 공부 시간(x), 시험 성적(y)을 뽑아내서 별도의 리스트에 저장한다.
x = [i[0] for i in data]
y = [i[1] for i in data]

for i in range(len(data)):
    print('공부 시간: {}, 시험 성적: {}'.format(x[i], y[i]))

공부 시간: 2, 시험 성적: 81
공부 시간: 4, 시험 성적: 93
공부 시간: 6, 시험 성적: 91
공부 시간: 8, 시험 성적: 97


In [33]:
# Y = ax + b 수식에 a(기울기) b(y절편) 값을 대입해서 결과(예측 성적)를 계산하는 함수
def predict(x):
    return ab[0] * x + ab[1]

In [34]:
# 임의로 정한 기울기와 y절편으로 예측 성적을 계산한다.
predict_result = [] # 예측 성적을 저장할 빈 리스트
# 모든 공부 시간(x) 값을 한 번씩 대입해서 예측 성적 리스트(predict_result)에 추가한다.
for i in range(len(x)):
    predict_result.append(predict(x[i]))
    print('공부 시간: {}, 시험 성적: {}, 예측 성적 : {}'.format(x[i], y[i], predict_result[i]))

공부 시간: 2, 시험 성적: 81, 예측 성적 : 82
공부 시간: 4, 시험 성적: 93, 예측 성적 : 88
공부 시간: 6, 시험 성적: 91, 예측 성적 : 94
공부 시간: 8, 시험 성적: 97, 예측 성적 : 100


In [35]:
# 평균 제곱근 오차를 계산하는 함수
# rmse(예측 성적, 실제 성적) => 인수로 넘겨받는 데이터는 넘파이 배열 타입의 데이터이다. 일반적인 리스트는 반복문 통해 넣어야 한다. 
def rmse(predict_result, y):
    return np.sqrt(((predict_result - y) ** 2 ).mean())

In [36]:
# 평균 제곱근 오차를 계산할 데이터를 넘파이 배열로 만들어 평균 제곱근 오차를 계산하는 함수 rmse()를 호출하는 함수
def rmse_val(predict_result , y):
    return rmse(np.array(predict_result), np.array(y))

In [37]:
# 평균 제곱근 오차를 계산해서 출력한다.
rmse_result = rmse_val(predict_result, y)
print('RMSE: {}'.format(rmse_result))

RMSE: 3.3166247903554


기울기 2.1, y절편 82일 경우 RMSE는 3.535533905932736가 계산되는 것을 확인할 수 있다.  
기울기 3, y절편 76일 경우 RMSE는 3.3166247903554가 계산되는 것을 확인할 수 있다.  
이 오차를 줄이면서 새로운 선을 긋는 작업을 해야한다.   
이를 위해서는 기울기와 y절편을 적절히 조절하면서 오차의 변화를 살펴보고 오차가 최소가 되는 기울기와 y 절편을 찾아야 한다.  

기울기 2.3, y절편 79일 경우 RMSE는 2.880972058177584가 계산되는 것을 확인할 수 있다.  